<img src="./Untitled.png">

In [1]:
import pandas as pd

In [2]:
df = pd.DataFrame({
    'city':['City A','City B','City C','City D','City E','City F','City G','City H','City I','City J'],
    'days to stay':[3,4,5,2,3,1,6,8,1,7],
    'cost':[80,85,100,45,75,25,148,205,24,200]
})

<h1>BRUTE FORCE<h1>


In [3]:
sdf = df.sort_values(by=['days to stay','cost'])
sdf

city  days to stay  cost
8  City I             1    24
5  City F             1    25
3  City D             2    45
4  City E             3    75
0  City A             3    80
1  City B             4    85
2  City C             5   100
6  City G             6   148
9  City J             7   200
7  City H             8   205

In [4]:
table = []
for i in range(sdf.shape[0]):
    d = sdf.iloc[i][1]
    c = sdf.iloc[i][2]
    cy = []
    da = 14
    cy.append(sdf.iloc[i][0])
    for j in range(sdf.shape[0]):
        if i !=j :
            if da - sdf.iloc[j][1] > 0 :
                da -= sdf.iloc[j][1]               
                d += sdf.iloc[j][1]
                c += sdf.iloc[j][2]
                cy.append(sdf.iloc[j][0]) 
    if d > 14:
        d=0
        c=0
        cy.clear()    
    table.append([d,c,cy])
  
n=len(table)
for i in range(n-1,0,-1): 
    if table[i][0] == 0:
        table.pop(i)


for i in table[0][2]:
    print("เมืองที่ไป :",i)
print("จำนวนเมืองที่ไป :",len(table[0][2]))
print("จำนวนค่าใช้จ่าย :",table[0][1])
print("จำนวนวันที่เที่ยว :",table[0][0])
        


เมืองที่ไป : City I
เมืองที่ไป : City F
เมืองที่ไป : City D
เมืองที่ไป : City E
เมืองที่ไป : City A
เมืองที่ไป : City B
จำนวนเมืองที่ไป : 6
จำนวนค่าใช้จ่าย : 334
จำนวนวันที่เที่ยว : 14


<h5>จากการทดลอง<h5>
    ได้ทำการสร้างตารางขึ้นมาตามข้อมูลที่มี แล้ว ทำการเรียงลำดับจากน้อยไปมากของ days to stay และ cost แล้วจึงทำการหาผลลัพธ์แบบ Brute force 
    ผลลัพธ์ที่ได้คือ เมืองที่ท่องเที่ยวได้สูงสุดคือ 6 เมือง มีค่าใช่จ่ายทั้งหมด 334$ และใช้เวลาทั้งหมด 14 วัน 
    หากไม่ได้ทำการเรียงข้อมูล ผลลัพธ์ที่ได้จะมีความซับซ้อนมากขึ้น ต้องเขียนโค๊ดมากขึ้น
    

<h1>dynamic programming<h1>

In [5]:

n=df.shape[0]
wt = df.iloc
plan = 14
table = [[0 for x in range(plan+1)] for x in range(n+1)] 
for i in range(n+1): 
    for j in range(plan+1): 
        if i == 0 or j == 0: 
            table[i][j] = 0

        elif wt[i-1][1] <= j:
            table[i][j] = max(df.iloc[i-1][2] + table[i-1][j-wt[i-1][1]],  table[i-1][j]) 
        else: 
            table[i][j] = table[i-1][j]
print(table[n][plan])


380


In [6]:
tb = pd.DataFrame({
    '0':table[0],
    '1':table[1],
    '2':table[2],
    '3':table[3],
    '4':table[4],
    '5':table[5],
    '6':table[6],
    '7':table[7],
    '8':table[8],
    '9':table[9],
    '10':table[10],
    
})

tb.transpose()

0   1   2   3    4    5    6    7    8    9    10   11   12   13   14
0    0   0   0   0    0    0    0    0    0    0    0    0    0    0    0
1    0   0   0  80   80   80   80   80   80   80   80   80   80   80   80
2    0   0   0  80   85   85   85  165  165  165  165  165  165  165  165
3    0   0   0  80   85  100  100  165  180  185  185  185  265  265  265
4    0   0  45  80   85  125  130  165  180  210  225  230  265  265  310
5    0   0  45  80   85  125  155  165  200  210  240  255  285  300  310
6    0  25  45  80  105  125  155  180  200  225  240  265  285  310  325
7    0  25  45  80  105  125  155  180  200  228  253  273  303  328  348
8    0  25  45  80  105  125  155  180  205  230  253  285  310  330  360
9    0  25  49  80  105  129  155  180  205  230  254  285  310  334  360
10   0  25  49  80  105  129  155  200  225  249  280  305  329  355  380

จากการทดลอง
    โดยพบว่ามีความซับซ้อนมากกว่า Brute Force เพราะมีการใช้ค่าหลายตัวแปร และเก็บข้อมูลเป็นจำนวนมาก โดยผลลัพธ์ที่ได้คือ 380$ แต่ไม่สามารถทราบจำนวนเมืองหรือ ระยะเวลาที่ท่องเที่ยวได้

<h1>greedy technique<h1>

In [7]:
df.insert(3,"cost/day",[80/3,85/4,100/5,45/2,75/3,25/1,148/6,205/8,24/1,200/7])


In [8]:
cddf = df.sort_values(by=['cost/day'])
cddf

city  days to stay  cost   cost/day
2  City C             5   100  20.000000
1  City B             4    85  21.250000
3  City D             2    45  22.500000
8  City I             1    24  24.000000
6  City G             6   148  24.666667
4  City E             3    75  25.000000
5  City F             1    25  25.000000
7  City H             8   205  25.625000
0  City A             3    80  26.666667
9  City J             7   200  28.571429

In [9]:
day =0
n = cddf.shape[0]
index = 0
listindex = []
while True:
    day += cddf.iloc[index][1]
    listindex.append(index)
    index += 1
    if day == 11:
        index += 2
    if day == 14:
        break
    
costcddf = 0
for i in listindex:
    print(cddf.iloc[i][0])
    costcddf += cddf.iloc[i][2]
    
print("จำนวนเมืองที่ไป :",len(listindex))
print("จำนวนค่าใช้จ่าย :",costcddf)
print("จำนวนวันที่เที่ยว :",day)


City C
City B
City D
City E
จำนวนเมืองที่ไป : 4
จำนวนค่าใช้จ่าย : 305
จำนวนวันที่เที่ยว : 14


<h5>จากการทดลอง<h5>
    ได้ทำการหาอัตราส่วนของแต่ละเมือง โดยนำค่าท่องเที่ยวหารด้วยจำนวนวัน และ ทำการจัดเรียงลำดับของเมืองจากค่าใช้จ่ายน้อยไปค่าใช้จ่ายมาก
    โดยการใช้วิธี greedy จะไม่ใช้ผลลัพธ์ที่ดีที่สุดแต่เป็นผลลัพธ์ที่คุ้มค่ากับเงินที่สุดด และ ตรงตามเงื่อนไขที่ท่องเที่ยวเป็นเวลา 14 วัน โดยมีค่าใช้จ่ายอยู่ที 305$ ที่เที่ยว 4 เมือง โดยวิธีการคิดแบบ greedy สามารถหาได้ง่ายจากการคำนวณมือ แต่การเขียน code นั้นจะมีความซับซ้อนมากขึ้น

<h1>GA<h1>

In [10]:
from numpy import *
Distance =   array([
[9,12,15,6,9,3,18,24,3,21],
[12,16,20,8,12,4,24,32,4,28],
[15,20,25,10,15,5,30,40,5,35],
[6,8,10,4,6,2,12,16,2,14],
[9,12,15,6,9,3,18,24,3,21],
[3,4,5,2,3,1,6,8,1,7],
[18,24,30,12,18,6,36,48,6,42],
[24,32,40,16,24,8,48,64,8,56],
[3,4,5,2,3,1,6,8,1,7],
[21,28,35,14,21,7,42,56,7,49]])
Province = ['A','B','C','D','E','F','G','H','I','J']

In [11]:
Day = { 'A' : 3,'B' : 4,'C' : 5,'D' : 2,'E' : 3,'F' : 1,'G' : 6, 'H' : 8,'I' : 1, 'J' : 7}
Cost = { 'A' : 80,'B' : 85,'C' : 100,'D' : 45,'E' : 75,'F' : 25,'G' : 148,'H' : 205,'I' : 24, 'J' : 200}

In [12]:
Distance[Province.index('A'),Province.index('D')]

6

In [13]:
def decode(individual):
    sorted_individual = sorted(individual,reverse=True)
    sorted_index = list()
    for i in range(len(individual)):
        sorted_index.append(individual.index(sorted_individual[i]))
    return sorted_index

In [14]:
 decode([752, 242, 894, 817, 13, 467, 8, 449, 895, 419])

[8, 2, 3, 0, 5, 7, 9, 1, 4, 6]

In [15]:
chromosome1 = [146,306,55,694,544,267,822,611,868,418]
chromosome2 = [845, 964, 51,635,445,185,484,97,532,270]
s1 = decode(chromosome1)
print(s1)
print([Province[s1[i]] for i in range(len(s1))])
s2 = decode(chromosome2)
print(s2)
print([Province[s2[i]] for i in range(len(s2))])

[8, 6, 3, 7, 4, 9, 1, 5, 0, 2]
['I', 'G', 'D', 'H', 'E', 'J', 'B', 'F', 'A', 'C']
[1, 0, 3, 8, 6, 4, 9, 5, 7, 2]
['B', 'A', 'D', 'I', 'G', 'E', 'J', 'F', 'H', 'C']


In [16]:
newchromosome1 = [146,306,55,694,544,267,822,611,868,418]
newchromosome2 = [845, 964, 51,635,445,185,484,97,532,270]
s1 = decode(newchromosome1)
print(s1)
print([Province[s1[i]] for i in range(len(s1))])
s2 = decode(newchromosome2)
print(s2)
print([Province[s2[i]] for i in range(len(s2))])

[8, 6, 3, 7, 4, 9, 1, 5, 0, 2]
['I', 'G', 'D', 'H', 'E', 'J', 'B', 'F', 'A', 'C']
[1, 0, 3, 8, 6, 4, 9, 5, 7, 2]
['B', 'A', 'D', 'I', 'G', 'E', 'J', 'F', 'H', 'C']


In [17]:
def generatePop(NumPop):
    population = list()
    x = 1
    while(x):
        x = 0
        for individual in range(NumPop):
            population.append(list())
            for gene in range(len(Province)):
                population[individual].append(random.randint(1,1000))
            for k in range(len(population[individual])):
                if population[individual].count(population[individual][k]) != 1:
                    population[individual][k] += 1
            for k in range(len(population[individual])):
                if population[individual].count(population[individual][k]) != 1:
                    population[individual][k] += 1
            for k in range(len(population[individual])):
                if population[individual].count(population[individual][k]) != 1:
                    x = 1
    return population

In [18]:
population = generatePop(10)

In [19]:
for p in range(len(population)):
    s = decode(population[p])
    print([Province[s[i]] for i in range(len(population[0]))])

['G', 'I', 'F', 'E', 'C', 'D', 'A', 'J', 'H', 'B']
['H', 'A', 'G', 'I', 'J', 'F', 'C', 'B', 'E', 'D']
['A', 'H', 'C', 'G', 'J', 'I', 'B', 'D', 'E', 'F']
['I', 'C', 'G', 'E', 'B', 'A', 'J', 'H', 'F', 'D']
['H', 'I', 'F', 'D', 'J', 'A', 'C', 'B', 'G', 'E']
['F', 'H', 'I', 'J', 'C', 'B', 'E', 'D', 'A', 'G']
['B', 'F', 'J', 'C', 'I', 'E', 'H', 'A', 'D', 'G']
['B', 'C', 'F', 'I', 'D', 'A', 'E', 'H', 'J', 'G']
['E', 'A', 'F', 'C', 'G', 'D', 'H', 'I', 'J', 'B']
['H', 'E', 'F', 'B', 'A', 'I', 'J', 'D', 'C', 'G']


In [20]:
def fitness1(individual):
    '''minus distance'''
    fitness_value = 0
    r = decode(individual)
    for i in range(len(individual)-4):
        #print('Distance from {0} to {1} = {2}'.format(Province[r[i%len(individual)]],Province[r[(i+1)%len(individual)]],Distance[r[i%len(individual)],r[(i+1)%len(individual)]]))
        #fitness_value += Distance[r[i%len(individual)],r[(i+1)%len(individual)]]
        #print(r)
        #print(individual)
        fitness_value += Distance[r[i%len(individual)],r[(i+1)%len(individual)]]
        #fitness_value += Distance4[r[i%len(individual)],r[(i+1)%len(individual)]]
        #print(fitness_value)
    #print('Total {}'.format(fitness_value))
    return fitness_value* -1

def fitness2(individual):
    '''Different of distance from max'''
    fitness_value = 0
    total = 0
    max_distance = sum([max(Distance[i]) for i in range(len(Distance))])
    r = decode(individual)
    P = [Province[r[i%len(individual)]] for i in range(len(individual))]
    D = [Distance[r[i%len(individual)],r[(i+1)%len(individual)]] for i in range(len(individual))]
    total = sum([Distance[r[i%len(individual)],r[(i+1)%len(individual)]] for i in range(len(individual))])
    
    fitness_value = sum([max_distance - Distance[r[i%len(individual)],r[(i+1)%len(individual)]] for i in range(len(individual))])
    #print('Max_distance'.format(max_distance))
    #print('Province {}'.format(P))
    #print('Distance {}'.format(D))
    #print('Total Distance {}'.format(total))
    #print('Fitness_value {}'.format(fitness_value))
    return fitness_value

def fitness3(individual):
    '''Power of different of distance from max'''
    fitness_value = 0
    total = 0
    max_distance = sum([max(Distance[i]) for i in range(len(Distance))])
    r = decode(individual)
    P = [Province[r[i%len(individual)]] for i in range(len(individual))]
    D = [Distance[r[i%len(individual)],r[(i+1)%len(individual)]] for i in range(len(individual))]
    total = sum([Distance[r[i%len(individual)],r[(i+1)%len(individual)]] for i in range(len(individual))])
    fitness_value = sum([(max_distance - Distance[r[i%len(individual)],r[(i+1)%len(individual)]])**2 for i in range(len(individual))])
    #print('Max_distance'.format(max_distance))
    #print('Province {}'.format(P))
    #print('Distance {}'.format(D))
    #print('Total Distance {}'.format(total))
    #print('Fitness_value {}'.format(fitness_value))
    return fitness_value

In [21]:
for i in range(len(population)):
    #print(population[i])
    print(fitness1(population[i]))

-41
-67
-147
-98
-61
-90
-78
-43
-75
-53


In [22]:
for i in range(len(population)):
    print(fitness2(population[i]))

3026
3079
3033
3034
3053
3074
3056
3011
3070
3045


In [23]:
for i in range(len(population)):
    print(fitness3(population[i]))

918224
948455
922143
922946
932727
945738
935072
909639
943256
929103


In [24]:
import numpy.random as npr
def roulette_wheel_selection(population):
    max = sum([fitness1(pop) for pop in population])
    selection_p = [fitness1(pop)/max for pop in population]
    select = npr.choice(len(population),p = selection_p)
    print('Select population[{0}]: {1} with probabily:{2:.2f} '.format(select,population[select],selection_p[select]))
    #print('Select population[{0}]: {1} with probabily:{2:.2f} '.format(select,decode(population[select]),selection_p[select]))
    return population[select]

In [25]:
roulette_wheel_selection(population)

Select population[5]: [238, 588, 832, 250, 487, 941, 13, 908, 906, 889] with probabily:0.12 


[238, 588, 832, 250, 487, 941, 13, 908, 906, 889]

In [26]:
def select_parent_population(population,k):
    selected_parents = list()
    for i in range(k):
        selected_parent = roulette_wheel_selection(population)
        selected_parents.append(selected_parent)
    return selected_parents

def make_unique(chromosome):
    for i in range(len(chromosome)-1):
        for j in range(i+1,len(chromosome)):
            if (chromosome[i] == chromosome[j]):
                chromosome[i] += chromosome.count(chromosome[i]) -1
                print('Change chromosome at position[{0}] to {1}'.format(i,chromosome[i]))
    return chromosome

def crossover(parent1,parent2):
    parent1 = list(parent1)
    parent2 = list(parent2)
    cp = random.randint(1,len(parent1))# no clone allowed
    # crossover point
    # cp = 0 [|0 1 2 3 4 5] - clone
    # cp = 1 [0|1 2 3 4 5]
    # cp = 2 [0 1|2 3 4 5]
    # cp = 3 [0 1 2|3 4 5]
    # cp = 4 [0 1 2 3|4 5]
    # cp = 5 [0 1 2 3 4|5]
    # cp = 6 [0 1 2 3 4 5|] - clone
    print('Crossover point: {}'.format(cp))
    print('Parent1: {0} x Parent2: {1}'.format(parent1,parent2))
    #print('Parent1: {0} x Parent2: {1}'.format(decode(parent1),decode(parent2)))
    offspring1,offspring2 = parent1[:cp],parent2[:cp]
    offspring1.extend(parent2[cp:])
    offspring2.extend(parent1[cp:])
    offspring1 = make_unique(offspring1)
    offspring2 = make_unique(offspring2)
    print('Offspring1: {0} , Offspring2: {1}'.format(offspring1,offspring2))
    #print('Offspring1: {0} , Offspring2: {1}'.format(decode(offspring1),decode(offspring2)))
    return offspring1,offspring2

def mutation(individual):
    '''Randomly swap position'''
    mutated_ind = list()
    mp1 = random.randint(0,len(individual))
    mp2 = random.randint(0,len(individual))
    mutated_ind[:] = individual[:]
    mutated_ind[mp1],mutated_ind[mp2] = mutated_ind[mp2],mutated_ind[mp1]
    print('Swap position {0} and {1} = Mutated: {2}'.format(mp1,mp2,mutated_ind))
    #print('Swap position {0} and {1} = Mutated: {2}'.format(mp1,mp2,decode(mutated_ind)))
    return mutated_ind

def breeding_offspring(population,num_cross,num_mutate):
    #Crossover operation
    print('------ Crossover -------')
    selected_parents = select_parent_population(population,num_cross)
    for i in range(int(num_cross/2)):
        offspring1, offspring2 = crossover(selected_parents[i*2],selected_parents[i*2+1])
        population.extend([offspring1,offspring2])
    #Mutantation operation
    print('------ Mutation -------')
    mutated_inds = list()
    for i in range(num_mutate):
        rand_ind = random.randint(0,len(population))
        print('Mutating Population[{0}]: {1}'.format(rand_ind,(population[rand_ind])))
        mutated_ind = mutation(population[rand_ind])
        mutated_inds.append(mutated_ind)
    population.extend(mutated_inds)
    print('-----------------------')

In [27]:
offspring1, offspring2 = crossover(population[0],population[1])

Crossover point: 1
Parent1: [305, 61, 488, 384, 490, 549, 901, 86, 900, 289] x Parent2: [948, 7, 187, 2, 4, 228, 868, 989, 453, 280]
Offspring1: [305, 7, 187, 2, 4, 228, 868, 989, 453, 280] , Offspring2: [948, 61, 488, 384, 490, 549, 901, 86, 900, 289]


In [28]:
mutated = mutation(population[0])

Swap position 6 and 0 = Mutated: [901, 61, 488, 384, 490, 549, 305, 86, 900, 289]


In [29]:
breeding_offspring(population,4,1)

------ Crossover -------
Select population[2]: [991, 540, 800, 353, 220, 176, 775, 937, 546, 669] with probabily:0.20 
Select population[2]: [991, 540, 800, 353, 220, 176, 775, 937, 546, 669] with probabily:0.20 
Select population[3]: [489, 598, 692, 117, 616, 352, 619, 458, 730, 462] with probabily:0.13 
Select population[2]: [991, 540, 800, 353, 220, 176, 775, 937, 546, 669] with probabily:0.20 
Crossover point: 8
Parent1: [991, 540, 800, 353, 220, 176, 775, 937, 546, 669] x Parent2: [991, 540, 800, 353, 220, 176, 775, 937, 546, 669]
Offspring1: [991, 540, 800, 353, 220, 176, 775, 937, 546, 669] , Offspring2: [991, 540, 800, 353, 220, 176, 775, 937, 546, 669]
Crossover point: 1
Parent1: [489, 598, 692, 117, 616, 352, 619, 458, 730, 462] x Parent2: [991, 540, 800, 353, 220, 176, 775, 937, 546, 669]
Offspring1: [489, 540, 800, 353, 220, 176, 775, 937, 546, 669] , Offspring2: [991, 598, 692, 117, 616, 352, 619, 458, 730, 462]
------ Mutation -------
Mutating Population[9]: [709, 780, 24

In [30]:
population
len(population)

15

In [31]:
def update_population(population,num_pop):
    for i in range(len(population)):
        print('population[{0}]: {1} fitness: {2}'.format(i,population[i],fitness1(population[i])))
        #print('population[{0}]: {1} fitness: {2}'.format(i,decode(population[i]),fitness1(population[i])))
        
    num_kill_pop = len(population) - num_pop
    evaluation = [fitness1(population[i]) for i in range(len(population))]
    for i in range(num_kill_pop):
        #if min(evaluation) < -14:
        kill_pop = evaluation.index(min(evaluation))
        #else:
             #kill_pop = evaluation.index(max(evaluation))
        print('kill {0} with fitness: {1}'.format(population[kill_pop],evaluation[kill_pop]))
        #print('kill {0} with fitness: {1}'.format(decode(population[kill_pop]),evaluation[kill_pop]))
        evaluation.remove(min(evaluation))
        population.remove(population[kill_pop])
    max_fitness_value = max(evaluation)
    max_fit_pop = population[evaluation.index(max(evaluation))]
    return max_fit_pop,max_fitness_value

In [32]:
max_fit_pop, max_fitness_value = update_population(population,10)

population[0]: [305, 61, 488, 384, 490, 549, 901, 86, 900, 289] fitness: -41
population[1]: [948, 7, 187, 2, 4, 228, 868, 989, 453, 280] fitness: -67
population[2]: [991, 540, 800, 353, 220, 176, 775, 937, 546, 669] fitness: -147
population[3]: [489, 598, 692, 117, 616, 352, 619, 458, 730, 462] fitness: -98
population[4]: [792, 402, 607, 829, 87, 931, 378, 971, 963, 819] fitness: -61
population[5]: [238, 588, 832, 250, 487, 941, 13, 908, 906, 889] fitness: -90
population[6]: [408, 965, 754, 407, 512, 888, 74, 420, 521, 808] fitness: -78
population[7]: [644, 992, 971, 711, 523, 798, 294, 461, 722, 382] fitness: -43
population[8]: [753, 8, 413, 232, 839, 593, 351, 167, 166, 146] fitness: -75
population[9]: [709, 780, 247, 355, 803, 798, 77, 893, 560, 395] fitness: -53
population[10]: [991, 540, 800, 353, 220, 176, 775, 937, 546, 669] fitness: -147
population[11]: [991, 540, 800, 353, 220, 176, 775, 937, 546, 669] fitness: -147
population[12]: [489, 540, 800, 353, 220, 176, 775, 937, 546,

In [33]:
population
len(population)

10

In [34]:
def genetic_algorithm():
    num_iteration = 100
    num_pop = 200
    num_cross = 6
    num_mutate = 2
    totalday = 0
    totalcost = 0
    # Generate Population Generation[0]
    population = generatePop(num_pop)
    # Loop for breeding and update population
    for iter in range(num_iteration):
        print('- Generation[{0}] -'.format(iter))
        for i in range(len(population)):
            print('population[{0}]: {1} fitness: {2}'.format(i,population[i],fitness1(population[i])))
            #print('population[{0}]: {1} fitness: {2}'.format(i,decode(population[i]),fitness1(population[i])))    
        breeding_offspring(population,num_cross,num_mutate)
        max_fit_pop,max_fitness_value = update_population(population,num_pop)

        # display result at Generation[i]
        s = decode(max_fit_pop)
        provinceList = [Province[s[i]] for i in range(len(population[0]))]
        total_distance = fitness1(max_fit_pop)*-1
        print('- End Generation[{0}]: Result:{1} with fitness: {2} -'.format(iter,max_fit_pop,max_fitness_value))
        #print('- End Generation[{0}]: Result:{1} with fitness: {2} -'.format(iter,decode(max_fit_pop),max_fitness_value))
        print('- Best path:{0} with fitness : {1} -'.format(provinceList,total_distance))
        totalday = 0
        totalcost = 0
        print("\n\n=========== Summary ==========")
        print('|\tNo.\t|\tCity\t|\tDays\t|\tCost\t|')
        for i in range (len(provinceList)-4):
            #print('- City : {0} stay with {1} days. Cost : {2} -'.format(provinceList[i],Day[provinceList[i]],Cost[provinceList[i]]))
            print('|\t{0}\t|\t {1} \t|\t  {2}  \t|\t{3}  \t|'.format(i+1,provinceList[i],Day[provinceList[i]],Cost[provinceList[i]]))
            totalday += Day[provinceList[i]] 
            totalcost += Cost[provinceList[i]]
        print('---------------------------------------------------------')
        print('Total days is : {0}'.format(totalday))
        print('Total cost is : {0}'.format(totalcost))
        print("============================")
        return totalcost


In [35]:
import time
start = time.time( )
genetic_algorithm()
end = time.time( )
elapsed = end - start
print('Time : {:.10f} s.'.format(elapsed))

- Generation[0] -
population[0]: [625, 5, 381, 653, 360, 606, 738, 459, 92, 657] fitness: -113
population[1]: [711, 968, 87, 900, 923, 784, 141, 185, 426, 459] fitness: -51
population[2]: [999, 694, 570, 935, 746, 483, 733, 428, 424, 354] fitness: -79
population[3]: [890, 20, 288, 473, 724, 701, 261, 591, 254, 480] fitness: -100
population[4]: [63, 660, 361, 723, 130, 590, 785, 232, 266, 757] fitness: -78
population[5]: [394, 652, 2, 401, 515, 829, 912, 979, 925, 377] fitness: -42
population[6]: [664, 848, 545, 614, 330, 210, 345, 663, 540, 173] fitness: -73
population[7]: [533, 62, 766, 412, 719, 814, 96, 929, 397, 102] fitness: -45
population[8]: [276, 975, 389, 164, 835, 126, 476, 952, 426, 760] fitness: -130
population[9]: [528, 968, 286, 715, 41, 6, 95, 288, 970, 654] fitness: -111
population[10]: [519, 532, 959, 413, 126, 112, 418, 118, 649, 708] fitness: -88
population[11]: [274, 60, 289, 947, 368, 442, 93, 818, 506, 165] fitness: -58
population[12]: [711, 939, 310, 106, 757, 4,

จากการทดลอง
    โดยวิธีการ GA ใช้เวลามากเนื่องจากมีข้อมูลให้ประมวลผลจำนวนมาก ผลลัพธ์ที่ได้ตรงกับวิธีแรก คือ brute force ที่มีค่าเดินทางที่เท่ากัน และ จำนวนเมืองที่เที่ยวที่เหมือนกัน
    วิธี GA มีความซับซ้อนมากที่สุดทำให้การทำงานนั้นล่าช้ามากขึ้น 